### ISSUE TITLE
Remove twisted.python.threadable's dependency on twisted.python.hook

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5860]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5860|
|Type|task|
|Created|2012-08-13 14:43:39Z|
|Branch|https://github.com/twisted/twisted/tree/unhook-threadable-5860|

`threadable.synchronize` is the only code in Twisted that depends on `hook`. Rewriting it to not use `hook` would mean less code to port to Python 3.

<details><summary>Searchable metadata</summary>

```
trac-id__5860 5860
type__task task
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_unhook_threadable_5860 branches/unhook-threadable-5860
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1344869019000000 1344869019000000
changetime__1346692729000000 1346692729000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [35477]) Branching to 'unhook-threadable-5860'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [35478]) Re-implement threadable.synchronize with a function wrapper instead of a hook

refs [#5860](https://github.com/twisted/twisted/issues/5860)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Removed the dependency and ported `threadable` to Python 3.  The astute reader will notice that `synchronize` is a class decorator.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/unhook-threadable-5860)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst set owner to @itamarst|
|-|-|

  1. You can't raise `unittest.FailTest` in Python 3. Since this only happens if the test fails, it just means the failure (which doesn't happen since the code is correct) is obscured. [#5885](https://github.com/twisted/twisted/issues/5885) will fix this, so I don't think this worth fixing.
  1. Not depending on reactor is good.
  1. I opened a ticket to improve testing of the synchronize decorator ([#5938](https://github.com/twisted/twisted/issues/5938)).

I note that Twisted cannot run without threading being available (since `twisted.python.context` uses it for example), so perhaps another ticket should be filed for simplifying the module. Not that we'll ever get around to it.

So, looks good. I will merge.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst set status to `closed`|
|-|-|

(In [35498]) Merge unhook-threadable-5860

Author: exarkun
Review: itamar
Fixes: [#5860](https://github.com/twisted/twisted/issues/5860)

Port twisted.python.threadable to Python 3.
